### PR TITLE
gnuplot based charting service Fixes #1419

### LINKED ISSUE #1419: charting service should be able to plot multiple lines and support numerical x axis
timeseries

### PR DESCRIPTION
**Pull request by hamidzr**
added a better charting support through writing a js friendly wrapper around gnuplot
- automatically sorted x axis
- smart padding
- smoothing option
- line types
- auto multiline support
- control x ticks
- time series
Also introducing options object as an input standard for optional and/or more advanced settings

requires #1522 #1512

### CODE REVIEW

File: src/server/rpc/procedures/chart/gnuPlotTest.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
+const gnuPlot = require('./node-gnuplot.js'),
```
Comment (brollb) [inline src/server/rpc/procedures/chart/gnuPlotTest.js]: If this is a testing file, can you move it to `test/`?

File: test/client/services.spec.js [hunk 2]
```diff
@@ -1,32 +1,67 @@
 /* globals expect, driver, MessageCreatorMorph, PushButtonMorph, NetsProcess*/
 describe('services', function() {
+
+
     before(function() {
         driver.selectCategory('network');
     });
 
-    describe('rpc inputs', function() {
+    describe.only('rpc inputs', function() {
```
Comment (brollb) [inline test/client/services.spec.js]: can you remove `.only`?

File: test/client/services.spec.js [hunk 3]
```diff
@@ -1,32 +1,67 @@
...
+            return serialized;
+        }
+
+        it('should serialize lists properly', function() {
+            var listInput = new List([new List([2,3]),new List([3,4])]);
+            var expectedOutput = 'arg[0][0]=2&arg[0][1]=3&arg[1][0]=3&arg[1][1]=4';
+            expect(serialize(listInput)).to.be(expectedOutput);
         });
 
+        it('should serialize [[0,1],[1,"null"],[2,null],[3],null,"null"]', function() {
```
Comment (brollb) [inline test/client/services.spec.js]: add a test, `.skip` or remove this

File: src/server/rpc/procedures/chart/gnuPlotTest.js (hunk 1 above)
Comment (hamidzr) [reply src/server/rpc/procedures/chart/gnuPlotTest.js]: it is not an automated test so where would be good place to put it?

**Comment by hamidzr** [comment]
>>  why are the qs changes on the client here also (they are in their own PR - are they required for this PR?)?

yes they are required, once you merge those in these should merge in no problem.

**Comment by brollb** [comment]
About the comment for removing the extra files: testing files can still be in the project - they just need to be moved into `test/`

File: test/server/rpc/procedures/chart.spec.js [hunk 4]
```diff
@@ -6,54 +6,7 @@ describe('chart', function() {
...
-        
-        it('test process one dataset', function() {
-            assert.deepEqual(chart._rpc._processData(dataset1,  'Age', 'bar'), supposedData1);
-        });
-    
-        it('test process multiple datasets', function() {
-            assert.deepEqual(chart._rpc._processMultipleData(dataset2,  ['1', '2'], 'bar'), supposedData2);
-        });
-    });
+
```
Comment (brollb) [inline test/server/rpc/procedures/chart.spec.js]: where are these tests?

File: test/client/debug.html [hunk 5]
```diff
@@ -0,0 +1,104 @@
+<html>
```
Comment (brollb) [inline test/client/debug.html]: why did you add this file? It looks like it was added elsewhere and changed...

File: src/server/rpc/procedures/dev/dev.js [hunk 6]
```diff
@@ -3,6 +3,6 @@ const rpcUtils = require('../utils/index');
 module.exports = {
     isSupported: () => process.env.ENV !== 'production',
     echo: function (argument){
-        return rpcUtils.jsonToSnapList(argument);
+        return argument;
```
Comment (brollb) [inline src/server/rpc/procedures/dev/dev.js]: Not sure why this changed... It looks like this is part of a dependent PR....

File: src/server/rpc/procedures/chart/node-gnuplot.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+const lineDefaults = {
+    title: '',
+    type: 'lines',
+    smoothing: undefined,
+    points: []
+};
+
+// opts defaults
+const optsDefautls = {
```
Comment (brollb) [inline src/server/rpc/procedures/chart/node-gnuplot.js]: `optsDefaults`

File: src/server/rpc/procedures/chart/node-gnuplot.js [hunk 8]
```diff
@@ -0,0 +1,101 @@
+/* eslint-disable quotes */
```
Comment (brollb) [inline src/server/rpc/procedures/chart/node-gnuplot.js]: why did you add this? I am not sure if this file justifies an exception to this rule...

File: test/server/rpc/procedures/chart.spec.js (hunk 4 above)
Comment (hamidzr) [reply test/server/rpc/procedures/chart.spec.js]: those were testing the custom methods that were used for the previous approach. they no longer apply

Comment (hamidzr) [reply test/server/rpc/procedures/chart.spec.js]: I'll add the interface tests once the argument names are fixed

File: src/server/rpc/procedures/dev/dev.js (hunk 6 above)
Comment (hamidzr) [reply src/server/rpc/procedures/dev/dev.js]: merged in from a dependent PR. After that one is merged and there is a merge conflict I'll take care of this one.

File: src/server/rpc/procedures/chart/node-gnuplot.js (hunk 8 above)
Comment (hamidzr) [reply src/server/rpc/procedures/chart/node-gnuplot.js]: because I'm using the single quotes inside the strings thus I can't use em to surround sentences with the same character, but the rule doesn't detect that.

**Comment by brollb** [comment]
- [ ] add gnuplot dependencies to Dockerfile
- [ ] improve error handling with unexpected input

File: test/client/services.spec.js [hunk 9]
```diff
@@ -1,64 +1,32 @@
...
-            var service = 'Dev',
-                rpcSignature = ['echo', ['arg']],
-                process = new NetsProcess();
-            process.getJSFromRPCDropdown = (service, rpc, params)=> params;
-            var serialized = process.getJSFromRPCStruct(service, rpcSignature, input);
-            return serialized;
-        }
-
         it('should serialize lists properly', function() {
+            var service = 'Dev',
```
Comment (brollb) [inline test/client/services.spec.js]: why is this changed?

**Comment by brollb** [comment]
Can you update the chart test in `test/client/rpc/`?

**Comment by hamidzr** [comment]
updating would be 'removing' here. Since the example is doing something similar

**Comment by brollb** [comment]
Actually, it might be better to move the example into the test since it is not very polished (no costumes or explanation around it explaining what it is doing).

**Comment by brollb** [comment]
Also, I tried to graph a bunch of temperature measurements:

image

but it failed on the server:

snap0017

**Comment by hamidzr** [comment]
> Also, I tried to graph a bunch of temperature measurements: but it failed on the server:

is your data valid?

**Comment by brollb** [comment]
Just updated the server screenshot to include the log for the initial invocation. Should be good data.

**Comment by hamidzr** [comment]
from the screenshot, it looks like you are trying to draw a one-dimensional chart. The service needs at least one line
``` [[x1,y1][x2,yw]..] ```

**Comment by brollb** [comment]
Can you fix the code climate issues?

**Comment by hamidzr** [comment]
done. the reset are fixed in another PR except one unused var due to using `arguments`

**Comment by brollb** [comment]
The y-axis can get messed up sometimes... I suspect there is a lexical sort rather than a numerical sort...

snap0020


snap0019
